### PR TITLE
[MDS-6206] Add support for custom permit condition categories

### PR DESCRIPTION
**Pull request by simensma-fresh**
## Objective 

[MDS-6206](https://bcmines.atlassian.net/browse/MDS-0000)

This PR adds functionality to allow for extracting, adding, editing, reordering, and deleting custom permit condition categories for Mines Act Permits. Also updated the permit condition extraction pipeline to add the conditions as is.

image

image

image

**Extra**
- Enabled versioning and soft deletion of categories to capture history
- Bumped up jest `maxWorkers` to `50%` <-- reduced the time to run Minespace tests on my laptop from 1min 10s to 45s
- Bumped version of actions/upload-artifact and actions/download-artifact in our Github Actions to v4 as v3 is being deprecated

### CODE REVIEW

File: cypress/realm-export.json [hunk 1]
```diff
@@ -759,7 +766,49 @@
...
+                "Chief Inspector",
+                "core_edit_historical_amendments",
+                "c_mds_health_safety",
+                "Data Cleanup Team",
+                "MDS Keycloak Realm Admin",
+                "core_executive_view",
+                "c_idir",
+                "cs_mds_investigations",
+                "cs_mds_permit_template_conditions",
+                "Administrative Users",
```
Comment (simensma-fresh) [inline cypress/realm-export.json]: Added a bunch of roles that the cypress user didn't have

File: services/core-api/app/api/mines/permits/permit_conditions/resources/permit_amendment_condition_category_resource.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+            permit_condition_category_code)
+        
+        if len(conditions) > 0:
+            raise BadRequest('Cannot delete a category that has conditions associated with it')
+
+        category.delete(commit=True)
+
+        return {'message': 'Permit condition category deleted successfully'}, 204
+
+    @requires_role_view_all
```
Comment (matbusby-fw) [inline services/core-api/app/api/mines/permits/permit_conditions/resources/permit_amendment_condition_category_resource.py]: Looks like you added `@requires_role_view_all` twice here

File: services/core-api/app/api/projects/project_summary/models/project_summary.py [hunk 3]
```diff
@@ -129,7 +136,8 @@ class ProjectSummary(SoftDeleteMixin, AuditMixin, Base):
     )
 
     municipality = db.relationship(
-        'Municipality', lazy='joined', foreign_keys=nearest_municipality_guid
+        'Municipality', lazy='joined', foreign_keys=nearest_municipality_guid,
```
Comment (matbusby-fw) [inline services/core-api/app/api/projects/project_summary/models/project_summary.py:139]: Ah, will this address that municipality warning/message that pops up in the logs?

Comment (simensma-fresh) [reply services/core-api/app/api/projects/project_summary/models/project_summary.py:139]: Yes!

File: services/common/src/components/common/ScrollSideMenu.tsx [hunk 4]
```diff
@@ -1,6 +1,6 @@
 import React, { FC, useEffect } from "react";
 import { useParams, useHistory, useLocation } from "react-router-dom";
-import { Anchor } from "antd";
+import { Anchor, Steps } from "antd";
```
Comment (matbusby-fw) [inline services/common/src/components/common/ScrollSideMenu.tsx]: Was this import intentional?  I don't see it being used.

File: services/common/src/components/common/ScrollSidePageWrapper.tsx [hunk 5]
```diff
@@ -8,16 +8,32 @@ interface ScrollSidePageWrapperProps {
...
+  extraItems?: ReactNode;
   headerHeight?: number;
+  view?: "default" | "steps" | "anchor";
 }
 
 export const coreHeaderHeight = 62; // match scss variable $header-height
 const msHeaderHeight = 80;
 
+/**
+ * A wrapper component that provides a side menu and a content area. The side menu can be fixed to the top of the page.
```
Comment (matbusby-fw) [inline services/common/src/components/common/ScrollSidePageWrapper.tsx:20]: I always really love that you add these descriptions.  I need to make more of a habit of doing that when appropriate.

File: services/common/src/redux/slices/permitConditionCategorySlice.ts [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        const params = new URLSearchParams();
+        if (payload.query) params.append("query", payload.query);
+        if (payload.exclude) payload.exclude.forEach((item) => params.append("exclude", item));
+        if (payload.limit) params.append("limit", payload.limit.toString());
+
+        const response = await CustomAxios({
+          errorToastMessage: "default",
+        }).get(`${ENVIRONMENT.apiUrl}/mines/permits/condition-category-codes?${params.toString()}`, headers);
+
+        console.log('Got response', response)
```
Comment (matbusby-fw) [inline services/common/src/redux/slices/permitConditionCategorySlice.ts]: Oops, missed a log

File: services/core-web/src/components/mine/Permit/PermitConditions.spec.tsx [hunk 7]
```diff
@@ -52,6 +54,56 @@ describe("PermitConditions", () => {
...
+    expect(stepInput).toBeInTheDocument();
+
+    const AddCategoryButton = screen.getByText("Add Category");
+    expect(AddCategoryButton).toBeInTheDocument();
+
+    AddCategoryButton.click();
+
+    await waitFor(() => { expect(screen.getAllByText("This is a required field")).toHaveLength(2); })
+
+    fireEvent.mouseDown(descriptionInput);
```
Comment (matbusby-fw) [inline services/core-web/src/components/mine/Permit/PermitConditions.spec.tsx:94]: I don't think it's super important, but the docs recommend using userEvent over fireEvent, as it better recreates the actual actions a user would be taking.

Something like:

const user = userEvent.setup()
await user.click(descriptionInput)

**Review by matbusby-fw — CHANGES_REQUESTED** [review]
Nice!  That's a biggie!  Just saw a couple of tiny things that can be removed, and one recommendation for changing how actions are fired in the one test.  Otherwise this is awesome!

File: services/core-web/src/components/mine/Permit/PermitConditions.spec.tsx (hunk 7 above)
Comment (simensma-fresh) [reply services/core-web/src/components/mine/Permit/PermitConditions.spec.tsx:94]: Yeah agreed. I just did a pass trying to set up the `user-event` package and hit a couple of issues getting the tests converted over. I might leave it for now as is as to not spend too much time, and keep that in mind for next time.

**Review by matbusby-fw — APPROVED** [review]
🦙

File: services/core-web/src/components/mine/Permit/PermitConditions.spec.tsx (hunk 7 above)
Comment (matbusby-fw) [reply services/core-web/src/components/mine/Permit/PermitConditions.spec.tsx:94]: Cool cool.  I think userEvent uses fireEvent under the hood and just also simulates some other behaviours as well, but functionally it shouldn't make too much difference.

**Review by sggerard — APPROVED** [review]
Looks great! 🐟

**Review by asinn134 — APPROVED** [review]

**Pull request merged**